// @imagekit/next/dist/client/index.js
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var javascript = require('@imagekit/javascript');
var NextImage = require('next/image');
var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var NextImage__default = /*#__PURE__*/_interopDefaultLegacy(NextImage);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

const ImageKitContext = React.createContext({});
/**
 * You can use the `ImageKitProvider` component to set the default values for all nested Image and Video components provided by the SDK.
 *
 * It accepts the following props:
 * - `urlEndpoint`: The URL endpoint for your ImageKit account. You can get this from the [ImageKit dashboard](https://imagekit.io/dashboard/url-endpoints).
 * - `transformationPosition`: The position of the transformation string in the URL. It can be either `path` or `query`. By default, it is set to `query`.
 *
 * @example
 * ```jsx
 * import { ImageKitProvider, Image } from "@imagekit/next";
 * <ImageKitProvider urlEndpoint="https://ik.imagekit.io/your_imagekit_id">
 *   <Image src="/default-image.jpg" />
 * </ImageKitProvider>
 * ```
 */
const ImageKitProvider = props => {
  const {
    urlEndpoint,
    transformationPosition
  } = props;
  return React__default["default"].createElement(ImageKitContext.Provider, {
    value: {
      urlEndpoint,
      transformationPosition
    }
  }, props.children);
};

/**
 * The Image component is a wrapper around the Next.js Image component. It supports all the features of the Next.js Image component, along with additional features provided by ImageKit.
 *
 * @example
 * ```jsx
 * import { Image } from "@imagekit/next";
 * <Image
 *  urlEndpoint="https://ik.imagekit.io/your_imagekit_id" // You can also set this in a parent ImageKitProvider component
 *  src="/default-image.jpg" // The path to the image in your ImageKit account
 *  alt="Default Image"
 *  width={500}
 *  height={500}
 *  transformation={[{ width: 500, height: 500 }]} // Add ImageKit transformations
 * />
 * ```
 */
const Image = props => {
  if (props.loader) {
    if (process.env.NODE_ENV !== "production") {
      console.warn('loader prop is ignored by ImageKit Image component.');
    }
  }
  const contextValues = React.useContext(ImageKitContext);
  // Its important to extract the ImageKit specific props from the props, so that we can use the rest of the props as is in the NextImage component
  const _a = Object.assign(Object.assign({}, contextValues), props // Override with props
    ),
    {
      transformation = [],
      unoptimized = false,
      quality,
      src = "",
      queryParameters,
      urlEndpoint,
      transformationPosition,
      responsive = true,
      loader
    } = _a,
    nonIKParams = __rest(_a, ["transformation", "unoptimized", "quality", "src", "queryParameters", "urlEndpoint", "transformationPosition", "responsive", "loader"]);
  if (!urlEndpoint || urlEndpoint.trim() === "") {
    if (process.env.NODE_ENV !== "production") {
      console.error("urlEndpoint is neither provided in this component nor in the ImageKitContext.");
    }
    return null;
  }
  const isAbsoluteURL = src.startsWith("http://") || src.startsWith("https://");
  // Do not mutate original transformation array from the props
  const finalTransformation = [...transformation];
  let propsTransformation = {};
  if (quality) {
    const parsedQuality = Number(quality);
    if (!isNaN(parsedQuality)) {
      propsTransformation.quality = parsedQuality;
    } else {
      if (process.env.NODE_ENV !== "production") {
        console.error('Invalid quality value, skipping transformation.');
      }
    }
  }
  // Return original file without any transformation or optimization to match the behavior of NextImage
  // Always keep src in the end
  if (unoptimized) {
    return React__default["default"].createElement(NextImage__default["default"], Object.assign({
      unoptimized: true
    }, nonIKParams, {
      src: javascript.buildSrc({
        urlEndpoint,
        src,
        queryParameters,
        transformationPosition,
        transformation: [{
          "raw": "orig-true"
        }]
      })
    }));
  }
  // User disabled responsive images
  if (!responsive) {
    return React__default["default"].createElement(NextImage__default["default"], Object.assign({
      unoptimized: true
    }, nonIKParams, {
      src: javascript.buildSrc({
        urlEndpoint,
        src,
        queryParameters,
        transformationPosition,
        // keep only user‑supplied transformations;no width, no crop
        transformation: [...finalTransformation, Object.assign({}, propsTransformation)]
      })
    }));
  }
  // Don't pass transformation just yet as loader will take care of it along with width for srcset generation as per Next.js logic.
  const finalSrcWithoutTransformation = javascript.buildSrc({
    urlEndpoint,
    queryParameters,
    src
  });
  return (
    // Always keep src in the end
    React__default["default"].createElement(NextImage__default["default"], Object.assign({
      loader: ({
        src,
        width
      }) => {
        // Since `src` inside loader is always absolute, `buildSrc` won't respect the `path` transformation position.
        // To fix this, if the original src prop isn't an absolute URL, remove `urlEndpoint` from this `src`
        // before calling `buildSrc`. Otherwise, leave it as is.
        const srcWithoutUrlEndpoint = !isAbsoluteURL ? src.replace(urlEndpoint, "") : src;
        const finalSrc = javascript.buildSrc({
          urlEndpoint,
          src: srcWithoutUrlEndpoint,
          transformationPosition,
          transformation: [...finalTransformation, Object.assign(Object.assign({}, propsTransformation), {
            width,
            crop: "at_max"
          })]
        });
        return finalSrc;
      }
    }, nonIKParams, {
      src: finalSrcWithoutTransformation
    }))
  );
};

/**
 * The Video component is a wrapper around the HTML video element. It supports all the features of the HTML video element, along with additional features provided by ImageKit.
 *
 * @example
 * ```jsx
 * import { Video } from "@imagekit/next";
 * <Video
 *  urlEndpoint="https://ik.imagekit.io/your_imagekit_id" // You can also set this in a parent ImageKitProvider component
 *  src="/default-video.mp4" // The path to the video in your ImageKit account
 *  controls
 *  width={500}
 *  height={500}
 *  transformation={[{ width: 500, height: 500 }]} // Add ImageKit transformations
 * />
 * ```
 *
 * @example
 * ```jsx
 * // With ref for programmatic control
 * const videoRef = useRef<HTMLVideoElement>(null);
 * <Video
 *  ref={videoRef}
 *  urlEndpoint="https://ik.imagekit.io/your_imagekit_id"
 *  src="/default-video.mp4"
 *  muted
 *  playsInline
 * />
 * ```
 */
const Video = React.forwardRef((props, ref) => {
  const contextValues = React.useContext(ImageKitContext);
  // Its important to extract the ImageKit specific props from the props, so that we can use the rest of the props as is in the video element
  const _a = Object.assign(Object.assign({}, contextValues), props // Override with props
    ),
    {
      transformation = [],
      src = "",
      queryParameters,
      urlEndpoint,
      transformationPosition
    } = _a,
    nonIKParams = __rest(_a, ["transformation", "src", "queryParameters", "urlEndpoint", "transformationPosition"]);
  if (!urlEndpoint || urlEndpoint.trim() === "") {
    if (process.env.NODE_ENV !== "production") {
      console.error("urlEndpoint is neither provided in this component nor in the ImageKitContext.");
    }
    return null;
  }
  const finalSrc = javascript.buildSrc({
    urlEndpoint,
    src,
    transformation: [...transformation],
    // Do not mutate original transformation array from the props
    queryParameters,
    transformationPosition
  });
  return React__default["default"].createElement("video", Object.assign({
    ref: ref
  }, nonIKParams, {
    src: finalSrc
  }));
});
Video.displayName = "Video";

Object.defineProperty(exports, 'ImageKitAbortError', {
    enumerable: true,
    get: function () { return javascript.ImageKitAbortError; }
});
Object.defineProperty(exports, 'ImageKitInvalidRequestError', {
    enumerable: true,
    get: function () { return javascript.ImageKitInvalidRequestError; }
});
Object.defineProperty(exports, 'ImageKitServerError', {
    enumerable: true,
    get: function () { return javascript.ImageKitServerError; }
});
Object.defineProperty(exports, 'ImageKitUploadNetworkError', {
    enumerable: true,
    get: function () { return javascript.ImageKitUploadNetworkError; }
});
Object.defineProperty(exports, 'buildSrc', {
    enumerable: true,
    get: function () { return javascript.buildSrc; }
});
Object.defineProperty(exports, 'buildTransformationString', {
    enumerable: true,
    get: function () { return javascript.buildTransformationString; }
});
Object.defineProperty(exports, 'getResponsiveImageAttributes', {
    enumerable: true,
    get: function () { return javascript.getResponsiveImageAttributes; }
});
Object.defineProperty(exports, 'upload', {
    enumerable: true,
    get: function () { return javascript.upload; }
});
exports.Image = Image;
exports.ImageKitContext = ImageKitContext;
exports.ImageKitProvider = ImageKitProvider;
exports.Video = Video;
